### DISCUSSION TITLE
Help: How to extend Pocketbase to include aggregate fields in resources

### DISCUSSION THREAD

**Discussion by tryffel** [pocketbase/pocketbase#265] [Q&A]
Hi and thank you for a great piece of software!

I was trying to figure out how to use aggregate values in resources, which I think is a core use case for relational databases. 
The documentation doesn't show examples for actually combining the manual queries into models and showing the data in REST api. 

Let's say I have a simple database of ```Posts``` and ```Comments```. For each Post I want to see the number of comments and timestamp of last comment. Thus ```comments_count``` and ```last_comment would``` be a virtual column for Post. How exactly would you implement this with Pocketbase? Right now I solved this with Sqlite triggers to automatically update additional aggregate columns in the target table, but I don't think this is a scalable solution.

I was thinking that ultimately a readonly datatype 'sql query' would be great for this purpose, e.g. ('count(documents) etc. Admins would be able to set virtual columns directly from the web ui. But ofc the MVP would be to simply extend the server and supply your own sql queries. 

Thank you for your help!

**Comment by ganigeorgiev** [comment]
Aggregations and computed/virtual columns are not supported out of the box and this is left to users to handle on their own.

Currently there isn't an easy to do what you want without using PocketBase as framework and querying the database manually.

Here is an example how this would look like (_I haven't tested it_):
```go
package main

import (
    "log"
    "net/http"

    "github.com/labstack/echo/v5"
    "github.com/pocketbase/pocketbase"
    "github.com/pocketbase/pocketbase/apis"
    "github.com/pocketbase/pocketbase/core"
    "github.com/pocketbase/pocketbase/tools/rest"
)

func main() {
    app := pocketbase.New()

    app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
        // GET /api/posts_comments_count?postIds=1&postIds=2&postIds=3&...
        e.Router.AddRoute(echo.Route{
            Method: http.MethodGet,
            Path:   "/api/posts_comments_count",
            Handler: func(c echo.Context) error {
                requestData := struct {
                    PostIds []string `form:"postIds" json:"postIds"`
                }{}

                if err := c.Bind(&requestData); err != nil {
                    return rest.NewBadRequestError("Failed to load the submitted data due to invalid formatting.", err)
                }

                result := []*struct {
                    PostId          string `db:"postId" json:"postId"`
                    TotalComments   int    `db:"totalComments" json:"totalComments"`
                    LastCommentDate string `db:"lastCommentDate" json:"lastCommentDate"`
                }{}

                queryErr := app.Dao().DB().
                    Select("postId", "COUNT(id) as totalComments", "MAX(created) as lastCommentDate").
                    From("comments").
                    GroupBy("postId").
                    All(&result)
                if queryErr != nil {
                    return rest.NewBadRequestError("Failed to fetch comments count.", queryErr)
                }

                return c.JSON(200, result)
            },
            // optionally enable an auth middleware
            // Middlewares: []echo.MiddlewareFunc{
            //     apis.RequireAdminOrUserAuth(),
            // },
        })

        return nil
    })

    if err := app.Start(); err != nil {
        log.Fatal(err)
    }
}
```

Your suggestion for "sql" type field is a nice one, but I'll have to think a little more on it because it may turn out not so easy as it seems.

**Reply by m-2k** [reply]
* `rest.NewBadRequestError` → `apis.NewBadRequestError`
* `apis.RequireAdminOrUserAuth()` not exists now
* `"github.com/pocketbase/pocketbase/tools/rest"` not need

**Reply by ganigeorgiev** [reply]
@m-2k This is an old post (pre v0.8) and a lot of things has changed since then. The alternative of `apis.RequireAdminOrUserAuth()` is `apis.RequireAdminOrRecordAuth()` (see also https://pocketbase.io/docs/go-routing/#builtin-middlewares).

**Reply by ganigeorgiev** [reply]
Also note that for aggregations you can now use [View collections](https://pocketbase.io/docs/collections/#view-collection).
